### DISCUSSION TITLE
i need your agreement to put your memory map in the software chirp

### DISCUSSION THREAD

**Discussion by joc2** [egzumer/uv-k5-firmware-custom#183] [Show and tell]
Hi
i want to put your memory map of the quansheng firmware you did in the chirp communication software.
i all ready contact chirp member (github project either) and they agree to implement it. But they ask to have your agreement to do it.

In the file setting, all the mapping memory is their, so it will take from their and make it full compatible in file driver in chirp
( key Menu, Backlight setting, etc... ) 

So, do you agree to take the memory mapping and transfert it in to chirp.

Thank's 

Jocelyn

**Comment by egzumer** [comment]
Sure, if you want, I'll appreciate help in that regard. I even started doing that. Here are my modifications if you can make use of that.
[uvk5.zip](https://github.com/egzumer/uv-k5-firmware-custom/files/13539445/uvk5.zip)

**Reply by joc2** [reply]
thank's, it's appreciated. i will let you know

**Comment by ErikS-web** [comment]
Will this be a setting UVK5-EGZUMER or general UVK5  in Chirp ?

**Reply by joc2** [reply]
shoud be something like UVK5-EGZUMER ... not sure. i will see with chirp team

**Comment by joc2** [comment]
Hi,
many of the menu are add or change for chirp  file  but still have some question
1 - i can't find the RXMODE menu in memory map,  i found the definition ( dual rx respond, cross band, main         tx  dual rx,main only )           in file menu.c i find a track of if but i dont see where it's store in memory map
     case MENU_TDR:               strcpy(String, gSubMenu_RXMode[gSubMenuSelection])     
2 - battery calibration setting in the lock menu of the radio at address 0x1f46 (2 byte long)     should we add capability to be change in chirp or leave only change in the radio ???
Let me know
Jocelyn

**Reply by egzumer** [reply]
1. The RXMode is a consolidation of Crossband and Dualwatch
   * Main only - CB off, DW off
   * Crossband - CB on, DW off
   * Dual rx respond - CB off, DW on
   * Main tx dual rx - CB on, DW on
1. Battery calibration should stay as a hidden menu option only

**Reply by joc2** [reply]
Hummm...  rxmode...a mix... ok,,,

the original CB and DW have 3 value each ( off, band A , band B ) = ( 0, 1, 2 ). by mixing those, i will have to think how to handle those in chirp for user uses....
so if
   CB = 0 or CB different then 0   ( 1 or 2 as value, give the same result )
 DW = 0 or DW different then 0  ( 1 or 2 as value, give the same result )

i think, i will have to remove the CB and DW  in chirp to the new display RXMODE.
a RXMode menu will be easier for user ...    then with the value select, i will write the right value to each register to give the result needed....
more complex that what i tought for this rxmode....

**Reply by egzumer** [reply]
Yeah I forgot about the upper/lower VFO selection. This is not used in the firmware at all. For me this is either on or off where:
OFF: 0
ON: 1 or 2

If the value is ON this is equal to address 0x0EAB (gEeprom.TX_VFO) +1 just for the compatibility sake

**Comment by egzumer** [comment]
I wonder, what did you do with stock QS options that are disabled in my build by default?

**Comment by joc2** [comment]
for the option that you not enable it in your compile the memory map are just not uses and skip that option in the reading example in testing.c  noaa_auto_scan at address 0xe73. not uses.  i leave the option in the chirp  . so if it. enable later. i will be programable . so in chirp all option can be programable either if the radio not uses it..that way it will be compatible in future version
what do you think???

**Reply by egzumer** [reply]
Yes, that is probably the best option. I don't know how that would look, but there could be some marking added next to options that are not available by default. Maybe a star, and a note that options marked may not be available. This is just a proposition, you don't have to add this, I leave this to for you to decide.

**Reply by joc2** [reply]
the chirp file can't know what is the option that is enable or not by reading the radio. it just get data of the memory map.
To do that with a star, it will need to implement a table for all option of the compilation for enable and desable ...
ouf... the memory map not done for that, it's a good idea but for implement it, it's a other game... i will think about it......

**Reply by egzumer** [reply]
Yes that would have be added manually, knowing which options are enabled and which are not. By default I mean the options that are present in the official release. For example there is no 1750Hz tone, no alarm, so choosing this option for configurable buttons will have no effect, the was already one question about it: https://github.com/egzumer/uv-k5-firmware-custom/issues/196

**Comment by joc2** [comment]
what is QS??

**Comment by egzumer** [comment]
QS - quansheng

**Reply by joc2** [reply]
ok :)

**Comment by joc2** [comment]
after thinking about your * idea.....
for alarm and 1750: 
In the version egzumer in chirp  they will a * in front of the option not available and a message to tell user this  might not be available...

in futur chirp , the text at arrow in picture number  will be:
1: Side key 1 short press, The option begin with a * might not be available in this version

2: None     Flashlight     Power     Monitor    Scan    VOX    *Alarm    FM radio    *Transmit 1750Hz,    Lock Keypad    Switch VFO    VFO/MR    Switch Demodul
i think it's can be done that way... for those 2....
for the other.. l will check what it's involve.... need to talk with chirp team... it will be availaible mid december
Jocelyn

**Reply by joc2** [reply]
ha the photo not show by email... i put it directely in github now..
![chirpmodif star](https://github.com/egzumer/uv-k5-firmware-custom/assets/56229329/e7b7103a-b356-4fbf-a19a-6c3d61f27681)

**Reply by egzumer** [reply]
I would suggest putting the disclaimer in the "Infos" tab. It will be a little bit hidden, but wouldn't need to include the disclaimer on every option. I don't know how good you are at python, this would need some more work, but we could do it as I suggested in my last post on the bottom, then it wouldn't need any stars or disclaimers.

**Reply by joc2** [reply]
the problem with info tab, it's that info tab not their if the developper view not selected and most user will not gone put developer mode ON and can't see it.....   so many question will pop up.....   what is * meanning .....    
![chirp no developer mode](https://github.com/egzumer/uv-k5-firmware-custom/assets/56229329/d1fc1092-df57-4d39-9527-e8a9a732aff2)

**Reply by egzumer** [reply]
Sorry, I meant "Driver information" in settings tab

**Reply by joc2** [reply]
can be a option... i will ask team chirp what they do with all kind of other radio option to do something similar... i will let you know...

**Comment by ErikS-web** [comment]
If you are adjusting...
Also consider adjusting the frequency-limits checked by Chirp below 50 and above 600

![Error below 50 Mhz](https://github.com/egzumer/uv-k5-firmware-custom/assets/148579604/b1b3829c-9544-430e-8f26-3c49e6078db7)

**Comment by egzumer** [comment]
Frequency table should be as follows
> BANDS = {
        0: [18.0, 108.0],
        1: [108.0, 136.9999],
        2: [137.0, 173.9999],
        3: [174.0, 349.9999],
        4: [350.0, 399.9999],
        5: [400.0, 469.9999],
        6: [470.0, 1300.0]
        }

Also I got an idea, I could write to the EEPROM all options that have been built into the firmware. This would then be accessible in chirp.

BTW. Where will you publish the driver code?

**Reply by joc2** [reply]
the guy that did quansheng in team chirp will contact me it a week, is to busy for now ...  so will be later than that
nothing publish until contact him...

**Reply by egzumer** [reply]
It would be good thing to publish this driver in a repository on github eather way. That way changes can be tracked, it could be updated in the future, tested and so on. I can create a separate repository and make you a collaborator there. What do you think. Or will you start your own repo?

**Reply by joc2** [reply]
it' in .., 
i think that can be view in chirp on setting tab driver information,
need to implement in this tab...
that was a great idea to take the compiler option see in chirp soon......  :)   

#seekto 0x1FF0:
struct {
u8 ENABLE_FMRADIO:1, // bit 0
   ENABLE_NOAA:1, // bit 1
   ENABLE_VOICE:1, // bit 2
   ENABLE_VOX:1, // bit 3
   ENABLE_ALARM:1, // bit 4
   ENABLE_TX1750:1, // bit 5
   ENABLE_PWRON_PASSWORD:1, // bit 6
   ENABLE_DTMF_CALLING:1 // bit 7
} Compiler_Option_1

#seekto 0x1FF1:
struct {
u8 ENABLE_FLASHLIGHT:1, // bit 0
   ENABLE_WIDE_RX:1, // bit 1
   ENABLE_BYP_RAW_DEMODULATORS:1, // bit 2
   ENABLE_BLMIN_TMP_OFF:1, // bit 3
   NU_COMPILER_OPTION_4:1, // bit 4
   NU_COMPILER_OPTION_5:1, // bit 5
   NU_COMPILER_OPTION_6:1, // bit 6
   NU_COMPILER_OPTION_7:1 // bit 7
} Compiler_Option_2

**Comment by egzumer** [comment]
I added a function to write build options as bits in memory:

0x1FF0:
   * 0 - ENABLE_FMRADIO
   * 1 - ENABLE_NOAA
   * 2 - ENABLE_VOICE
   * 3 - ENABLE_VOX
   * 4 - ENABLE_ALARM
   * 5 - ENABLE_TX1750
   * 6 - ENABLE_PWRON_PASSWORD
   * 7 - ENABLE_DTMF_CALLING
 
0x1FF1:
   * 0 - ENABLE_FLASHLIGHT
   * 1 - ENABLE_WIDE_RX
   * 2 - ENABLE_BYP_RAW_DEMODULATORS
   * 3 - ENABLE_BLMIN_TMP_OFF

**Reply by joc2** [reply]
ok, i will add it. in the file,  :) 

to be sure, 1ff1 : if bit0 is = to 1, this is enable flashlight, else not implement in this version of firmware
if you modify it on the fly, check something with beep of the keypad... 
i was testing few antenna with different frequency 130 to 157 and 350 to 470 for the swr and while changing the frequency once in a while the keypad beep stop working, i change the frequency and the beep keypad came back .... realy strange.. this happen 2 time while testing 6 type of antenne... minor bug... :)

**Reply by egzumer** [reply]
Yes that is correct. bit0==1 means the FW was built with `ENABLE_FLASHLIGHT:= 1`

**Reply by egzumer** [reply]
> keypad beep stop working, i change the frequency and the beep keypad came back

I have no idea what that might be

**Comment by joc2** [comment]
>If you are adjusting...
> Also consider adjusting the frequency-limits checked by Chirp below 50 and above 600

 good point :)

**Comment by joc2** [comment]
> If you are adjusting...
> Also consider adjusting the frequency-limits checked by Chirp below 50 and above 600

 change as be done by ezgumer in the file he star to modify it and i am continuing it :) ...
it was 76 to 108

**Reply by joc2** [reply]
forget the it was 76 to 108 in my comment, it a cut a paste  that was their....    that the fm radio range 76mhz to 108mhz

**Comment by joc2** [comment]
when it reproduce, i will take care how to do it....    and let you know.
    Le jeudi 7 décembre 2023 à 19 h 36 min 40 s HNE, egzumer ***@***.***> a écrit :  
 
 



keypad beep stop working, i change the frequency and the beep keypad came back


I have no idea what that might be

**Comment by joc2** [comment]
contact me direct :  ***@***.*** if you want.

    Le jeudi 7 décembre 2023 à 19 h 17 min 32 s HNE, egzumer ***@***.***> a écrit :  
 
 


Use the github discussions page to communicate, emails make a mess, and you seem not to read some messages.
#183

**Reply by joc2** [reply]
what , it hiden my email address....  wouaw...

**Reply by egzumer** [reply]
here is fine

**Comment by joc2** [comment]
why not add frequency in memory... 😎

**Comment by joc2** [comment]
i saw that chirp not show the version  complet it cut the 9 in 0.19...  found that buffer was to small... will be fix....
Envoyé à partir de Yahoo Courriel sur Android

**Comment by joc2** [comment]
> It would be good thing to publish this driver in a repository on github eather way. That way changes can be tracked, it could be updated in the future, tested and so on. I can create a separate repository and make you a collaborator there. What do you think. Or will you start your own repo?

maybe...  but i want get contact of the chirp guy that first did the software for the radio, he told me he will be check with me mid december , i dont want make him feeling bad...  that's him who create it and allow to implement in chirp.
that's why i am waiting to get answers from him soon. i hope you understand.

**Reply by egzumer** [reply]
Well I don't understand. I don't know why would he feel bad. This is an open source and can be used freely. If it was published there would be a possibility of testing and 3rd party contribution. Testing is highly desired BEFORE the driver is added to CHIRP. Also it could be used right now as an external module, before it is incorporated into CHIRP.

**Comment by joc2** [comment]
> Yeah I forgot about the upper/lower VFO selection. This is not used in the firmware at all. For me this is either on or off where:
> OFF: 0
> ON: 1 or 2
>
> If the value is ON this is equal to address 0x0EAB (gEeprom.TX_VFO) +1

 ok i will check about it

**Comment by egzumer** [comment]
I would like to add two more options:
0x0EA1 - S0 s-meter level [dBm], default -130
0x0EA2 - S9 s-meter level [dBm], default -76

defaults should be shown if 0xFF is in the memory

these values in reality are always negative and should be presented with a minus, user input should probably be converted to a negative value, or there should be an error if positive values are entered, but in the memory there is an unsigned value. The allowed range of values is not that important I think, but this is 1 byte value so 0-254 (255 means undefined, default value is used)

**Comment by joc2** [comment]
ok, i will check to add this a those address.in chirp it will add  a minus signe  will check soon...
but the address 0xea1 is all ready in use in chirp ...???
#seekto 0xea0;u8 keypad_tone;u8 language;

    Le vendredi 8 décembre 2023 à 20 h 26 min 16 s HNE, egzumer ***@***.***> a écrit :  
 
 


I would like to add two more options:
0x0EA1 - S0 s-meter level [dBm], default -130
0x0EA2 - S9 s-meter level [dBm], default -76

defaults should be shown if 0xFF is in the memory

these values in reality are always negative and should be presented with a minus, user input should probably be converted to a negative value, or there should be an error if positive values are entered, but in the memory there is an unsigned value. The allowed range of values is not that important I think, but this is 1 byte value so 0-254 (255 means undefined, default value is used)

**Reply by egzumer** [reply]
No this is not used yet. But it is an easy mod. Will add that soon.

**Comment by joc2** [comment]
the guy contact me this after noon and he will check it.....

all modification in the file i made, i add  with a comment ( #joc +  comment ad it... )  this is for debug and fast way to to find modif from the original file that you did and with what jacek lipkowsk put in the original chirpwith winmerge program.
her is the file where i am now with it. nothing testedi try to compile it but the compiler give me error in the library of python in the main chirp. that i do not change at all...   




    Le vendredi 8 décembre 2023 à 20 h 33 min 22 s HNE, egzumer ***@***.***> a écrit :  
 
 


Well I don't understand. I don't know why would he feel bad. This is an open source and can be used freely. If it was published there would be a possibility of testing and 3rd party contribution. Testing is highly desired BEFORE the driver is added to CHIRP. Also it could be used right now as an external module, before it is incorporated into CHIRP.

**Reply by egzumer** [reply]
You don't need to compile anything for testing. You can load the driver as a module in chirp. You turn on the developer mode, then you go to file -> load external module

**Comment by joc2** [comment]
ok, this is a thing that i dont try.... that ... way. nice...

    Le vendredi 8 décembre 2023 à 21 h 46 min 32 s HNE, egzumer ***@***.***> a écrit :  
 
 


You don't need to compile anything for testing. You can load the driver as a module in chirp. You turn on the developer mode, then you go to file -> load external module

**Comment by joc2** [comment]
ok i was thinking # is a comment line as the begin of the file but it's not  it's // i am changing it now

    Le vendredi 8 décembre 2023 à 21 h 46 min 32 s HNE, egzumer ***@***.***> a écrit :  
 
 


You don't need to compile anything for testing. You can load the driver as a module in chirp. You turn on the developer mode, then you go to file -> load external module

**Comment by joc2** [comment]
forget about the file i send to you... a working one will be done soon...i am correcting all menu.... this is the long press M bouton
 
    Le vendredi 8 décembre 2023 à 21 h 46 min 32 s HNE, egzumer ***@***.***> a écrit :  
 
 


You don't need to compile anything for testing. You can load the driver as a module in chirp. You turn on the developer mode, then you go to file -> load external module

**Comment by joc2** [comment]
i found out the comment when reading in chirp reading with a strange behaviour....i gone remove all ...  
many menu working now...
i made a big mistake a the begining....  my error...;(
    Le vendredi 8 décembre 2023 à 21 h 46 min 32 s HNE, egzumer ***@***.***> a écrit :  
 
 


You don't need to compile anything for testing. You can load the driver as a module in chirp. You turn on the developer mode, then you go to file -> load external module

**Comment by egzumer** [comment]
Push changes to `devel` branch of https://github.com/egzumer/uvk5-chirp-driver

**Comment by joc2** [comment]
ok, it will be their by monday.. i am out of the house for 2 days.
    Le samedi 9 décembre 2023 à 11 h 09 min 19 s HNE, egzumer ***@***.***> a écrit :  
 
 


Push changes to devel branch of https://github.com/egzumer/uvk5-chirp-driver

**Comment by joc2** [comment]
file is their on github... not finish bu working( rxmode mix with dualwatch and crossband.   need to be tested all feature,    try it, by reading data from the radio, and see what the result in chirp.
like compiler option, s0,s9, etc...
Jocelyn
    Le samedi 9 décembre 2023 à 11 h 09 min 19 s HNE, egzumer ***@***.***> a écrit :  
 
 


Push changes to devel branch of https://github.com/egzumer/uvk5-chirp-driver

**Comment by ErikS-web** [comment]
Nice work. 
1 bug detected.
Menu 32. MICBar is ON in radio, but not V in Chirp

and works OK for 
Enter a frequentie 27 Mhz without Chirp error. And actually loaded in Radio.

**Comment by ErikS-web** [comment]
When saving settings, the filename had (egzumer). Suggest to remove the () and place 1 -

Not this format for filename
Quansheng_UV-K5 (egzumer)_20231211.img

Suggest this format for filename
Quansheng_UV-K5 - egzumer_20231211.img

See line 657 in source... 
I modified and tested locally

    MODEL = "UV-K5 - egzumer"

**Comment by joc2** [comment]
trouble find and fixthank's


    Le lundi 11 décembre 2023 à 15 h 49 min 45 s HNE, ErikS-web ***@***.***> a écrit :  
 
 


Menu 32. MICBar is ON in radio, but not V in Chirp

**Comment by joc2** [comment]
ok...
but it's not more clear with (egzumer) in the filename for the chirp uv-k5... 

Jocelyn
    Le lundi 11 décembre 2023 à 16 h 03 min 58 s HNE, ErikS-web ***@***.***> a écrit :  
 
 


When saving settings, the filename had (egzumer). Suggest to remove the () and place 1 -

Not this format for filename
Quansheng_UV-K5 (egzumer)_20231211.img

Suggest this format for filename
Quansheng_UV-K5 - egzumer_20231211.img

**Comment by joc2** [comment]
i just update the file to start adding text see on the radio to match chirp text., i saw you add a message box message. great, ok for desable from chirp. but a soon as your reatch this message box, it pop up at any change you try to make at any value on any option.
not sure what cause this...
Jocelyn

**Reply by egzumer** [reply]
It is because the validation routine is called every time there is a change on any setting, and for every possible option, not only the one that has been changed. Also because you have TX unlocked on your radio.
I'm trying to solve that.

**Comment by ErikS-web** [comment]
Download from radio

When I read with **first/previous version** I see, these values...
![First version](https://github.com/egzumer/uv-k5-firmware-custom/assets/148579604/e786157c-5a6a-4f90-872a-3859de32b771)

And with **new version dd 12-12-2023** I see
![second version dd 12-12-2023](https://github.com/egzumer/uv-k5-firmware-custom/assets/148579604/9554157d-c0c0-403d-b9c7-db498190fa4f)

**Reply by ErikS-web** [reply]
Aha detected why
![Reason - code change](https://github.com/egzumer/uv-k5-firmware-custom/assets/148579604/958361bd-7c03-4e36-9361-8b537bcac8d1)

**Comment by joc2** [comment]
yes, egzumer correct my mistake... one is 130 and the other is 76
by the way, i am in the modification to make the setting option the more same listing as on the radio... it will be their very soon, on the web i add the name show on the radio  in ( )  in chirp 
    Le mardi 12 décembre 2023 à 15 h 29 min 03 s HNE, ErikS-web ***@***.***> a écrit :  
 
 


Aha detected why
Reason.-.code.change.png (view on web)

**Reply by egzumer** [reply]
Please, please, please, use the `view it on GitHub` link and answer messages on the github webpage and not through emails. It is so much better that way. Look how the messages sent with email look, it's a mess.
